warehouse/query-core/src/main/java/datawave/query/jexl/visitors/RewriteNullFunctionsVisitor.java

```diff
@@ -0,0 +1,482 @@
+ * <ul>
+ * <li><code>filter:isNull(FIELD)</code> becomes <code>FIELD == null</code></li>
+ * <li><code>filter:isNull(F1 || F2)</code> becomes <code>F1 == null || F2 == null</code></li>
+ * <li><code>filter:isNotNull(FIELD)</code> becomes <code>!(FIELD == null)</code></li>
+ * <li><code>filter:isNotNull(F1 || F2)</code> becomes <code>!(F1 == null) || !(F2 == null)</code></li>
+ * </ul>
+ */
+public class RewriteNullFunctionsVisitor extends BaseVisitor {
+    
+    // used to rebuild flattened unions
+    private boolean rebuiltMultiFieldedFunction = false;
+    
+    private final static String IS_NOT_NULL = "isNotNull";
+    
+    private RewriteNullFunctionsVisitor() {}
+    
+    /**
+     *
+     * @param node
+     *            an arbitrary JexlNode
+     * @return the original query tree
+     */
+    @SuppressWarnings("unchecked")
+    public static <T extends JexlNode> T rewriteNullFunctions(JexlNode node) {
+        node.jjtAccept(new RewriteNullFunctionsVisitor(), null);
+        return (T) node;
+    }
+    
+    /**
+     * This is required to produce a flattened query tree
+     *
+     * @param node
+     *            an ASTOrNode
+     * @param data
+     *            an object
+     * @return the same node
+     */
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        node.childrenAccept(this, data);
+        
+        if (rebuiltMultiFieldedFunction) {
+            List<JexlNode> children = new ArrayList<>();
+            for (JexlNode child : JexlNodes.children(node)) {
+                JexlNode deref = JexlASTHelper.dereference(child);
+                if (deref instanceof ASTOrNode) {
+                    Collections.addAll(children, JexlNodes.children(deref));
+                } else {
+                    children.add(child);
+                }
+            }
+            JexlNodes.children(node, children.toArray(new JexlNode[0]));
+            rebuiltMultiFieldedFunction = false;
+        }
+        
+        return data;
+    }
+    
+    /**
+     * Rewrite <code>filter:isNull</code> and <code>filter:isNotNull</code> functions to the more efficient <code>FIELD == null</code> or
+     * <code>!(FIELD == null)</code>
+     *
+     * @param node
+     *            a function node
+     * @param data
+     *            an object
+     * @return the rewritten node, or the original node if no rewrite was required
+     */
+    @Override
+    public Object visit(ASTFunctionNode node, Object data) {
+        
+        FunctionJexlNodeVisitor visitor = new FunctionJexlNodeVisitor();
+        node.jjtAccept(visitor, null);
+        
+        if (visitor.namespace().equals(EVAL_PHASE_FUNCTION_NAMESPACE)) {
+            if (visitor.name().equals(IS_NULL)) {
+                JexlNode rewritten = rewriteFilterFunction(visitor, false);
+                JexlNodes.replaceChild(node.jjtGetParent(), node, rewritten);
```

Comment:
check that rewritten is not null?

Reply:
Good catch, I'm going to rewrite this whole block to be a little more efficient

Reply:
Should be done